### PR TITLE
MM_ANI_interpolation

### PR DESCRIPTION
**Pull request by dominicrufa**
first working script to interpolate small molecule forcefield from MM to ANI whilst retaining MM interactions intermolecularly.

### CODE REVIEW

**Comment by maxentile** [comment]
Looks good so far! Could you please convert the class definitions etc. to a `.py` file (e.g. using [nbconvert](https://nbconvert.readthedocs.io/en/stable/usage.html#convert-script)), so I can make inline comments?

**Comment by maxentile** [comment]
Excellent -- thanks!

**Review by dominicrufa — COMMENTED** [review]
as of right now, there _appears_ to be a problem with the thermostat (specifically, if attempting to propagate NOT in the overdamped regime, then the kinetic energy appears to stabilize at roughly different value than what it should be, given a maxwell_boltzmann distribution on velocities); however, i added a metropolization step as well as a recorder for the shadow work, so the correct particle weights can be easily recapitulated


if someone can find what i am doing wrong in the `simulate_baoab` function, i'll give you a cookie

File: execute.py
Comment (maxentile) [inline execute.py]: What does this line do? (gamma has unit 1/t_unit)

File: execute.py
Comment (maxentile) [inline execute.py]: Are you confident this function is correct? Hard for me to check all the reshaping and slicing by eye

File: execute.py
Comment (maxentile) [inline execute.py]: Signature should be something like V(x, v), as the force depends on x

File: execute.py
Comment (maxentile) [inline execute.py]: Could you please annotate the intent of the reshape / slicing logic in this line?

**Review by maxentile — COMMENTED** [review]
Unclear about intent of the collision-rate unit conversion, didn't carefully review the mass-matrix slicing/reshaping logic but it looked potentially error-prone

**Comment by maxentile** [comment]
>if someone can find what i am doing wrong in the simulate_baoab function, i'll give you a cookie

Took another look and flagged a couple suspicious bits

File: execute.py
Comment (dominicrufa) [reply execute.py]: make gamma unitless?

Comment (maxentile) [reply execute.py]: Okay!

The gamma-dependent term that appears in an exponential (`gamma * dt`) will already be unitless even when gamma and dt are unit'd.

(Throughout, I think units are being removed a bit "preemptively", not sure I follow the intent of this.)

**Comment by maxentile** [comment]
To help isolate the problem, I would like to rule out that there's an issue with the generalized mass-matrix slicing / reshaping.

Could we temporarily replace the general mass matrix with a single mass scalar (equivalent to a constant diagonal mass matrix) to check if the issues you describe persist?

**Comment by dominicrufa** [comment]
> To help isolate the problem, I would like to rule out that there's an issue with the generalized mass-matrix slicing / reshaping.
> 
> Could we temporarily replace the general mass matrix with a single mass scalar (equivalent to a constant diagonal mass matrix) to check if the issues you describe persist?

yeah, i've already tried using a diagonal mass matrix.  nothing changes

**Comment by maxentile** [comment]
>yeah, i've already tried using a diagonal mass matrix. nothing changes

If the result using a constant diagonal mass matrix `m = np.eye(n)` is identical to the result using a mass scalar `m = 1.0`, then that narrows it down!

**Comment by dominicrufa** [comment]
alright, i computed the incremental velocities two ways (one using the single row vector of masses with slicing/nasty stuff and the other using a diagonal matrix

**Comment by dominicrufa** [comment]
it looks like regardless of whether i use the old way of doing matrix operations with the mass matrix or simply make an NxN diagonal matrix, the kinetic energy of the incremental velocity is the same (i.e. the velocity update is the same)

File: execute.py
Comment (maxentile) [inline execute.py]: Does this need to be prefaced with `v = `?

Comment (dominicrufa) [reply execute.py]: no, it is purely for the sake of logging...if the script is run, it will print a `stupid` ke from the incremental velocity update and the ke of the `real` incremental velocity...eyeballing them gives the exact same ke...

Comment (maxentile) [reply execute.py]: Okay! I'm not 100% sure what's being tested by the new definition -- I had meant to suggest trying to simplify to a single scalar mass for the purpose of debugging.

I think I had misunderstood a lot here. I had read the code assuming mistakenly that mass_matrix was a square matrix, but just realized it is defined as an (n_particles, 1) vector. (So each appearance of `mass_matrix[0,:][:,None]` could be replaced with `mass_matrix.T` if you want.) Even better, could we rename this to `masses` and have it be of shape `(n_particles, )` instead?

Comment (dominicrufa) [reply execute.py]: i can't take `mass_matrix.T` without losing the mass units, so i use `mass_matrix[0,:][:,None]`...but all of this is consistent...based on the last test and the two logger statement reduced incremental kinetic energies, i am less than convinced that the problem is here...

Comment (maxentile) [reply execute.py]: Okay! Just a readability thing. ("mass matrix" usually refers to a square matrix, so I didn't realize it was a vector until I read the point where it was defined -- if there are just n_particles numbers could use the name "masses" or similar)

**Comment by maxentile** [comment]
Could we make a list of the remaining hypotheses for what may be causing the behavior you describe?

File: execute.py
Comment (dominicrufa) [reply execute.py]: i had originally named it mass matrix because i was convinced i was going to use an nxn diagonal matrix to define the masses and do algebra in the integrator, but performing matrix multiplication with it proved a little slow, so i switched it to a vector and left the name.  regardless, if you look through the logger comments, yo u can see the loop that is running the VRORV integrations between incrementing the lambda value.  the entire reason i thought that there was a bug here somewhere is because i noticed that the kinetic energy of the system seems to decrease rather dramatically and then level out somewhere toward the end of the annealing process if i am _not_ in the overdamped regime. it may be worthwhile to look at the logger comments that are spat out of the `anneal` method.  there is also a nonzero chance that there isn't a bug and i'm wrong

**Comment by maxentile** [comment]
General observations that might help simplify things: there are two separate tasks that we're trying to do here, but their implementations are currently a bit entangled in `InterpolAIS`:
1. Interpolate MM to ANI intramolecular energies / forces
2. Use this interpolated model to compute free energies, either by computing (a) instantaneous works, or (b) annealed works with some nonequilibrium process ((b1) one such nonequilibrium process might include underdamped langevin dynamics without a metropolis correction).

I think that the object that does the interpolation should ideally provide just two public functions: `potential_energy(x, lam)` and `force(x, lam)`. These can then be reused in various sampling and free energy estimation approaches as needed. Specifically, I think the `InterpolAIS` class also doesn't need to know the temperature/pressure (or other information about what thermodynamic state it will be used for), doesn't need to know anything about particle masses, and doesn't necessarily need to have history-accumulating logic (or logic for saving trajectories to disk). I see these as all somewhat separate concerns.

From the other side, I think that the implementation of a free energy estimator or approximate sampler doesn't necessarily need to know all of the details of how `potential_energy` or `force` are implemented, and doesn't need to live inside the `InterpolAIS` class. For example, Langevin only needs to interact with `force` as a black-box.

I would vote to separate these two tasks, if possible.

File: execute.py
Comment (maxentile) [reply execute.py]: Okay! I took a brief look at the log file but will need more time to understand what I'm looking at -- so far just noticed that KE by end of trajectory is much lower than the Maxwell-Boltzmann sampled KE...

**Comment by maxentile** [comment]
To simplify, I would propose to split the InterpolAIS class into something like:
- QMModel(residue), with methods energy(coordinates), force(coordinates). (An ANIModel would inherit from this, and its constructor would only need to look at the elements in the residue.)
- IntramolecularInterpolator(mm_system, qm_residue, QMModel), with methods energy(coordinates, lam), force(coordinates, lam)
- Standard samplers, estimators, etc. that interact with energy(coordinates, lam), force(coordinates, lam) functions.

If we want to swap out different QM-surrogate models (e.g. when ANI-2 comes out), we don't have to re-implement the samplers or interpolation logic. If we want to swap out samplers or free energy estimators, the QM/MM interpolation class doesn't have to contain new implementations for each (as long as they only interact with energies and forces, which I think covers most that we would use here @dominicrufa?).

File: complete.py
Comment (maxentile) [inline complete.py:5]: I haven't reviewed these yet, but we should review these!

File: execute.py
Comment (hannahbrucemacdonald) [inline execute.py]: INFO 🙃

File: execute.py
Comment (hannahbrucemacdonald) [inline execute.py]: Are these works and shadow-works?
would you mind documenting them pls

File: complete.py
Comment (maxentile) [inline complete.py:237]: docstring needs to be updated to describe "F" step

File: complete.py
Comment (maxentile) [inline complete.py:443]: is the iteration counter intended to be a float or an int?

File: execute.py
Comment (dominicrufa) [reply execute.py]: let me delete this file; it's `complete.py` that is the one i use; im sorry

File: complete.py
Comment (maxentile) [inline complete.py:411]: this convention is off by one from what I would expect -- is this typical? want to recover simple importance sampling when the number of intermediates is zero.

File: complete.py
Comment (maxentile) [inline complete.py:665]: can remove commented out code block if unused or move to a `demos` folder or similar if used

File: complete.py
Comment (hannahbrucemacdonald) [inline complete.py]: Could this function be called something different? I would expect `_calculate_energy` and `calculate_energy` to return the same thing maybe?

File: complete.py
Comment (hannahbrucemacdonald) [inline complete.py:407]: Is this a safe way to pass stuff in to the function? Is there a reason that it's not `_before_integration(particle_state, n_iterations,**kwargs)` ?

File: complete.py
Comment (hannahbrucemacdonald) [inline complete.py:442]: As before, I'm not sure why this is written like this?

File: complete.py
Comment (hannahbrucemacdonald) [inline complete.py:443]: Can iteration be an int?

**Review by maxentile — COMMENTED** [review]
Left a couple low-level comments -- I will need to revisit more carefully for possible off-by-one errors involving modified forces or importance weight increments. I wasn't able to check confidently in this review pass because the increment logic is a bit distributed.

If @dominicrufa could write a little algorithm-box pseudocode description of the intent, that would be very helpful to me. (And also helpful for the write-up I think!) Something at the level of abstraction of "at time t, I compute forces using F(x_{t-1} ; lambda_t) take a step to produce a sample x_t ~ p_t, and then compute an importance weight q_{t+1}(x_t)/q_t(x_t)" or similar, including any intermediate variables used to mix MM and ANI forces and energies. That would help check the implementation against the intent.

I think it would also be good to add a test to confirm that this neq implementation is consistent with simple importance sampling in the case of no intermediates. (Do MD to sample a snapshot from state 0 and compute one importance weight for perturbing from state 0 to state 1.) In this test I think the current implementation would produce two incremental importance weights (the first always 0.0), and I'm not sure, but I think in that test this code may be doing MD in state 1? (but then not using it in computing the importance weight?) It's probably the case that these off-by-one conventions all cancel out, but it's just hard for me to check.

File: complete.py
Comment (maxentile) [inline complete.py]: should this be "/" or "-" ?

File: complete.py
Comment (maxentile) [inline complete.py:311]: defining "p" as reduced potential threw me for a loop -- pretty ingrained to think of "p" as a normalized density or similar

File: complete.py
Comment (maxentile) [inline complete.py]: should "work_incremental" be the r.h.s. minus something?

File: complete.py
Comment (maxentile) [inline complete.py:322]: excellent, thanks, this clarifies!

File: complete.py
Comment (maxentile) [inline complete.py:325]: excellent, thanks, this note further clarifies!

File: complete.py
Comment (dominicrufa) [reply complete.py:407]: yes; i don't want to break the `super` class's API (defined in `coddiwomple.openmm.propagators`); in order to make the `Propagator.apply` method amenable to generalized SMC (not necessarily AIS) i need to allow the most flexibility in the `_before_integration` and `_during_integration` as possible, so i pass general args and kwargs

File: complete.py
Comment (dominicrufa) [reply complete.py:442]: see above

File: complete.py
Comment (dominicrufa) [reply complete.py:443]: and `CustomIntegrator` converts these to floats anyway

Comment (dominicrufa) [reply complete.py:443]: intended to be an int, but it's always treated consistently

File: complete.py
Comment (dominicrufa) [reply complete.py]: should be  '-' (it is computed properly inside the code); will change

File: complete.py
Comment (dominicrufa) [reply complete.py:311]: i am very inconsistent; it is treated like a reduced potential; can change

File: complete.py
Comment (wiederm) [inline complete.py]: purely cosmetic: that line can be deleted

File: complete.py
Comment (wiederm) [inline complete.py]: we still might want to assert that statment

File: complete.py
Comment (wiederm) [inline complete.py:161]: at some point this function could be deleted and simply replaced in code with 
`energy_in_hartree = self.model((self.species, coordinates)).energies`

File: complete.py
Comment (wiederm) [inline complete.py:320]: ah! took me a while to realize that the difference between ante_perturbation_potential and the first part of evaluate_work_incremental changed in its value for lambda! But very nice explanation!

File: complete.py
Comment (wiederm) [inline complete.py:9]: I am not really familiar with openmmtools -- just to clarify:
- you are using ThermoDynamicState as a way to compute the reduced_potential with respect to the ensemble?
- you are using SamplerState as a way to index/slice the topology and coordinates?

File: complete.py
Comment (wiederm) [inline complete.py:406]: are these private methods defined in the super class? are they called there?

File: complete.py
Comment (wiederm) [inline complete.py:460]: I am not sure what happens here -- could you explain this call a bit?

File: complete.py
Comment (wiederm) [inline complete.py:644]: I think this is not used subsequently

**Review by wiederm — APPROVED** [review]
Very nice work from @dominicrufa -- I think we can merge this!
There are some things I would appreciate your comment, but I think I will be able to figure this out when I am running this locally.

I think @maxentile made a good point about tests. I can start writing some after the pull request has been merged.

File: complete.py
Comment (dominicrufa) [reply complete.py:9]: i use `ThermodynamicState.reduced_potential` method on a sampler state as a convention to compute the reduced potential; the sampler state has to be updated from the context to update the box vectors and potential energy to the sampler state

Comment (dominicrufa) [reply complete.py:9]: sampler state holds positions, velcoities, and box vectors

File: complete.py
Comment (dominicrufa) [reply complete.py:406]: yes; the code looks a little janky because i didn't want to break the higher level API of the super `Propagator`

File: complete.py
Comment (dominicrufa) [reply complete.py:460]: since we don't use the context's `f` variable in the `V` step of the integration scheme, we have to create a `modified_f` that is a `perDofVariable` in the integrator which contains the force vector of the mm systems _augmented_ by the force matrix of the ani system. `_compute_hybrid_forces` creates this hybrid force matrix given the positions of the full system, and the value of lambda

Comment (dominicrufa) [reply complete.py:460]: then the modified force is passed to the integrator in the `setPerDofVariableByName`

**Comment by dominicrufa** [comment]
since we don't have any tests at the moment, i am going to merge in an hour. voice objections before then, please

File: complete.py
Comment (wiederm) [reply complete.py:9]: thanks for clarifying!

File: complete.py
Comment (wiederm) [reply complete.py:460]: thanks for clarifying! My confusion came from the setPerDofVariableByName; I was simply not sure why this gets the modified force, but now that makes sense!

**Pull request merged**
